File: android/core/src/main/kotlin/org/hdhropen/kit/networking/APIClient.kt
package org.hdhropen.kit.networking

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.hdhropen.kit.models.*
import org.hdhropen.kit.utilities.Log
import java.io.IOException
import java.util.concurrent.TimeUnit

class InMemoryCookieJar : CookieJar {
    private val cookieStore = mutableMapOf<String, MutableList<Cookie>>()

    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        val host = url.host
        val currentCookies = cookieStore.getOrPut(host) { mutableListOf() }
        cookies.forEach { newCookie ->
            currentCookies.removeAll { it.name == newCookie.name }
            currentCookies.add(newCookie)
        }
    }

    override fun loadForRequest(url: HttpUrl): List<Cookie> {
        val host = url.host
        val currentCookies = cookieStore[host] ?: return emptyList()
        val now = System.currentTimeMillis()
        currentCookies.removeAll { it.expiresAt < now }
        return currentCookies
    }

    fun clear() {
        cookieStore.clear()
    }
}

@Serializable
data class HLSSessionResponse(
    @SerialName("session_id")
    val sessionId: String,
    @SerialName("playlist_url")
    val playlistUrl: String
)

@Serializable
private data class LoginBody(
    val pin: String? = null,
    @SerialName("token_name")
    val tokenName: String? = null
)

@Serializable
private data class RecordingStreamHLSBody(
    val url: String,
    @SerialName("recording_id")
    val recordingId: String? = null,
    val start: Double? = null,
    @SerialName("audio_index")
    val audioIndex: Int? = null,
    val provider: String? = null,
    @SerialName("for_cast")
    val forCast: Boolean = false
)

class APIClient(
    var baseURL: String,
    val cookieJar: InMemoryCookieJar = InMemoryCookieJar(),
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .cookieJar(cookieJar)
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
) {
    var bearerToken: String? = null
    var deviceId: String? = null

    val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
        coerceInputValues = true
    }

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    // MARK: - Core HTTP Engine

    suspend fun requestRaw(
        path: String,
        method: String = "GET",
        jsonBody: String? = null,
        headers: Map<String, String> = emptyMap()
    ): ByteArray = withContext(Dispatchers.IO) {
        val fullUrl = if (path.startsWith("http://") || path.startsWith("https://")) {
            path
        } else {
            val base = baseURL.trimEnd('/')
            val cleanPath = if (path.startsWith("/")) path else "/$path"
            "$base$cleanPath"
        }

        val requestBuilder = Request.Builder().url(fullUrl)

        bearerToken?.let { token ->
            requestBuilder.header("Authorization", "Bearer $token")
        }

        deviceId?.let { id ->
            requestBuilder.header("X-Device-Id", id)
        }

        headers.forEach { (k, v) ->
            requestBuilder.header(k, v)
        }

        when (method.uppercase()) {
            "GET" -> requestBuilder.get()
            "POST" -> {
                val body = jsonBody?.toRequestBody(jsonMediaType) ?: "".toRequestBody(jsonMediaType)
                requestBuilder.post(body)
            }
            "DELETE" -> requestBuilder.delete()
            "PATCH" -> {
                val body = jsonBody?.toRequestBody(jsonMediaType) ?: "".toRequestBody(jsonMediaType)
                requestBuilder.patch(body)
            }
            "PUT" -> {
                val body = jsonBody?.toRequestBody(jsonMediaType) ?: "".toRequestBody(jsonMediaType)
                requestBuilder.put(body)
            }
        }

        val response: Response
        try {
            response = httpClient.newCall(requestBuilder.build()).execute()
        } catch (e: IOException) {
            throw APIError.NetworkError(e.localizedMessage ?: "Network error")
        }

        val bytes = response.body?.bytes() ?: ByteArray(0)

        if (!response.isSuccessful) {
            val detailMessage = extractErrorDetail(bytes) ?: "Request failed with status code ${response.code}"
            when (response.code) {
                401 -> throw APIError.Unauthorized(detailMessage)
                403 -> throw APIError.Forbidden
                404 -> throw APIError.NotFound(detailMessage)
                429 -> throw APIError.LockedOut(detailMessage)
                else -> throw APIError.ServerError(response.code, detailMessage)
            }
        }

        bytes
    }

    suspend inline fun <reified T> request(
        path: String,
        method: String = "GET",
        body: Any? = null,
        headers: Map<String, String> = emptyMap()
    ): T = withContext(Dispatchers.IO) {
        val jsonBody = body?.let {
            when (it) {
                is String -> it
                else -> json.encodeToString(it)
            }
        }
        val data = requestRaw(path, method, jsonBody, headers)
        val text = String(data, Charsets.UTF_8)
        try {
            json.decodeFromString<T>(text)
        } catch (e: Exception) {
            Log.network.error("JSON Decode error for $path: ${e.localizedMessage}")
            throw APIError.DecodingError(e.localizedMessage ?: "Decoding error")
        }
    }

    private fun extractErrorDetail(bytes: ByteArray): String? {
        if (bytes.isEmpty()) return null
        return try {
            val text = String(bytes, Charsets.UTF_8)
            val jsonObject = json.parseToJsonElement(text).jsonObject
            jsonObject["detail"]?.jsonPrimitive?.content
        } catch (e: Exception) {
            null
        }
    }

    // MARK: - Guide APIs

    suspend fun getChannels(): HDHomeRunChannelsResponse =
        request(APIEndpoints.guideChannels())

    suspend fun getGuide(): List<HDHomeRunFullGuideChannel> =
        request(APIEndpoints.guide())

    suspend fun refreshGuide() {
        requestRaw(APIEndpoints.refreshGuide(), method = "POST")
    }

    // MARK: - DVR APIs

    suspend fun getDvrInfo(): HDHomeRunDvrInfo =
        request(APIEndpoints.dvrInfo())

    suspend fun listRecordings(): List<HDHomeRunRecording> =
        request(APIEndpoints.recordings())

    suspend fun deleteRecording(id: String) {
        requestRaw(APIEndpoints.deleteRecording(id), method = "DELETE")
    }

    suspend fun getRecordingDetail(
        url: String,
        recordingId: String,
        start: Double? = null,
        recordEnd: Double? = null,
        provider: String? = null
    ): HDHomeRunRecordingDetail =
        request(APIEndpoints.recordingDetail(url, recordingId, start, recordEnd, provider))

    suspend fun listRecordingRules(): List<HDHomeRunRecordingRule> =
        request(APIEndpoints.recordingRules())

    suspend fun addRecordingRule(payload: AddRecordingRulePayload): List<HDHomeRunRecordingRule> =
        request(APIEndpoints.recordingRules(), method = "POST", body = json.encodeToString(payload))

    suspend fun deleteRecordingRule(id: String): List<HDHomeRunRecordingRule> =
        request(APIEndpoints.deleteRecordingRule(id), method = "DELETE")

    // MARK: - Live Watch APIs

    suspend fun startWatch(channelNumber: String): HDHomeRunRecording? {
        return try {
            val resp: HDHomeRunRecording = request(APIEndpoints.startWatch(channelNumber), method = "POST")
            if (resp.recordingId == null && resp.sessionId == null) null else resp
        } catch (e: Exception) {
            null
        }
    }

    suspend fun heartbeatWatch(sessionId: String) {
        requestRaw(APIEndpoints.heartbeatWatch(sessionId), method = "POST")
    }

    suspend fun stopWatch(sessionId: String) {
        requestRaw(APIEndpoints.stopWatch(sessionId), method = "POST")
    }

    suspend fun promoteWatch(sessionId: String, options: Map<String, JsonElement>? = null): HDHomeRunRecording {
        val bodyStr = options?.let { json.encodeToString(it) }
        return request(APIEndpoints.promoteWatch(sessionId), method = "POST", body = bodyStr)
    }

    // MARK: - HLS Packaging APIs

    suspend fun createChannelHLSSession(
        channelNumber: String,
        forCast: Boolean = false,
        audioIndex: Int? = null
    ): HDHomeRunRecording =
        request(APIEndpoints.hlsChannelSession(channelNumber, forCast, audioIndex), method = "POST")

    suspend fun createRecordingHLSSession(
        url: String,
        recordingId: String? = null,
        start: Double? = null,
        audioIndex: Int? = null,
        provider: String? = null,
        forCast: Boolean = false
    ): HLSSessionResponse {
        val body = RecordingStreamHLSBody(url, recordingId, start, audioIndex, provider, forCast)
        return request(APIEndpoints.hlsRecordingSession(), method = "POST", body = json.encodeToString(body))
    }

    suspend fun stopHLSSession(sessionId: String) {
        requestRaw(APIEndpoints.stopHLSSession(sessionId), method = "POST")
    }

    // MARK: - Tuner APIs

    suspend fun getTunerStatus(): List<HDHomeRunTuner> =
        request(APIEndpoints.tunerStatus())

    suspend fun getTunerInfo(): HDHomeRunTunerInfo =
        request(APIEndpoints.tunerInfo())

    // MARK: - Auth & Profile APIs

    suspend fun listProfiles(): List<UserProfile> =
        request(APIEndpoints.users())

    suspend fun login(userId: String, pin: String?, tokenName: String?): CurrentUser {
        val body = LoginBody(pin, tokenName)
        return request(APIEndpoints.loginUser(userId), method = "POST", body = json.encodeToString(body))
    }

    suspend fun logout() {
        requestRaw(APIEndpoints.logoutUser(), method = "POST")
    }

    suspend fun getCurrentUser(): CurrentUser =
        request(APIEndpoints.currentUser())

    suspend fun getUserPreferences(): UserPreferences =
        request(APIEndpoints.userPreferences())

    suspend fun updateUserPreferences(prefs: UserPreferences): UserPreferences =
        request(APIEndpoints.userPreferences(), method = "PATCH", body = json.encodeToString(prefs))

    // MARK: - Device APIs

    suspend fun registerDevice(): DeviceRegisterResult {
        val result: DeviceRegisterResult = request(APIEndpoints.registerDevice(), method = "POST")
        deviceId = result.id
        return result
    }

    suspend fun getCurrentDevice(): DeviceInfo =
        request(APIEndpoints.currentDevice())

    suspend fun listDevices(): List<DeviceListEntry> =
        request(APIEndpoints.listDevices())

    suspend fun deleteDevice(id: String) {
        requestRaw(APIEndpoints.deleteDevice(id), method = "DELETE")
    }

    // MARK: - Settings APIs

    suspend fun getSettings(): AppSettings =
        request(APIEndpoints.settings())

    suspend fun getTranscodePresets(): List<HDHomeRunTranscodePreset> =
        request(APIEndpoints.transcodePresets())

    suspend fun getHWAccelDiagnostics(): HWAccelDiagnostics =
        request(APIEndpoints.hwaccelDiagnostics())

    suspend fun listNetworkIntegrations(): List<NetworkIntegration> =
        request(APIEndpoints.networkIntegrations())

    suspend fun getNetworkIntegration(type: String): NetworkIntegration =
        request(APIEndpoints.networkIntegration(type))

    suspend fun updateNetworkIntegration(type: String, settings: Map<String, JsonElement>): NetworkIntegration =
        request(APIEndpoints.networkIntegration(type), method = "PATCH", body = json.encodeToString(settings))

    suspend fun fetchRawString(url: String): String = withContext(Dispatchers.IO) {
        val data = requestRaw(url)
        String(data, Charsets.UTF_8)
    }
}

File: apple/HDHROpenKit/Sources/HDHROpenKit/Networking/APIClient.swift
import Foundation

public struct HLSSessionResponse: Decodable, Sendable {
    public let sessionId: String
    public let playlistUrl: String

    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case playlistUrl = "playlist_url"
    }
}

public actor APIClient {
    public var baseURL: URL
    var bearerToken: String?
    var deviceId: String?
    let session: URLSession

    let jsonDecoder = JSONDecoder()

    let jsonEncoder = JSONEncoder()

    public init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    public func setBaseURL(_ url: URL) {
        baseURL = url
    }

    public func setBearerToken(_ token: String?) {
        bearerToken = token
    }

    public func currentBearerToken() -> String? {
        bearerToken
    }

    public func setDeviceId(_ id: String?) {
        deviceId = id
    }

    public func currentDeviceId() -> String? {
        deviceId
    }

    // MARK: - Core HTTP Engine

    public func request<T: Decodable>(
        path: String,
        method: String = "GET",
        body: (any Encodable)? = nil,
        headers: [String: String] = [:]
    ) async throws -> T {
        let data = try await requestRaw(path: path, method: method, body: body, headers: headers)
        do {
            return try jsonDecoder.decode(T.self, from: data)
        } catch {
            Log.network.error("JSON Decode error for \(path): \(error.localizedDescription)")
            throw APIError.decodingError(error.localizedDescription)
        }
    }

    public func requestRaw(
        path: String,
        method: String = "GET",
        body: (any Encodable)? = nil,
        headers: [String: String] = [:]
    ) async throws -> Data {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw APIError.invalidURL
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = method

        if let token = bearerToken {
            urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let devId = deviceId {
            urlRequest.setValue(devId, forHTTPHeaderField: "X-Device-Id")
        }

        for (k, v) in headers {
            urlRequest.setValue(v, forHTTPHeaderField: k)
        }

        if let body {
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try jsonEncoder.encode(body)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch {
            throw APIError.networkError(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.networkError("Invalid server response")
        }

        if httpResponse.statusCode >= 400 {
            let detailMessage = extractErrorDetail(from: data) ?? "Request failed with status code \(httpResponse.statusCode)"
            if httpResponse.statusCode == 401 {
                throw APIError.unauthorized(detailMessage)
            } else if httpResponse.statusCode == 403 {
                throw APIError.forbidden
            } else if httpResponse.statusCode == 404 {
                throw APIError.notFound(detailMessage)
            } else if httpResponse.statusCode == 429 {
                throw APIError.lockedOut(detailMessage)
            } else {
                throw APIError.serverError(statusCode: httpResponse.statusCode, message: detailMessage)
            }
        }

        return data
    }

    private func extractErrorDetail(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            if let raw = String(data: data, encoding: .utf8), !raw.isEmpty {
                Log.network.error("Non-JSON error body: \(raw, privacy: .public)")
            }
            return nil
        }
        if let detail = json["detail"] as? String {
            return detail
        }
        if let detailObject = json["detail"] as? [String: Any], let message = detailObject["message"] as? String {
            return message
        }
        if let error = json["error"] as? String {
            return error
        }
        Log.network.error("Unrecognized error body shape: \(String(describing: json), privacy: .public)")
        return nil
    }

    // MARK: - Guide APIs

    public func getChannels() async throws -> HDHomeRunChannelsResponse {
        try await request(path: APIEndpoints.guideChannels())
    }

    public func getGuide(start: Double? = nil, end: Double? = nil) async throws -> [HDHomeRunFullGuideChannel] {
        try await request(path: APIEndpoints.guide(start: start, end: end))
    }

    public func refreshGuide() async throws {
        _ = try await requestRaw(path: APIEndpoints.refreshGuide(), method: "POST")
    }

    // MARK: - DVR APIs

    public func getDvrInfo() async throws -> HDHomeRunDvrInfo {
        try await request(path: APIEndpoints.dvrInfo())
    }

    public func listRecordings() async throws -> [HDHomeRunRecording] {
        try await request(path: APIEndpoints.recordings())
    }

    public func deleteRecording(id: String) async throws {
        _ = try await requestRaw(path: APIEndpoints.deleteRecording(id), method: "DELETE")
    }

    public func getRecordingDetail(url: String, recordingId: String, start: Double? = nil, recordEnd: Double? = nil) async throws -> HDHomeRunRecordingDetail {
        try await request(path: APIEndpoints.recordingDetail(url: url, recordingId: recordingId, start: start, recordEnd: recordEnd))
    }

    public func listRecordingRules() async throws -> [HDHomeRunRecordingRule] {
        try await request(path: APIEndpoints.recordingRules())
    }

    public func addRecordingRule(payload: AddRecordingRulePayload) async throws -> [HDHomeRunRecordingRule] {
        try await request(path: APIEndpoints.recordingRules(), method: "POST", body: payload)
    }

    public func updateRecordingRule(id: String, payload: AddRecordingRulePayload) async throws -> [HDHomeRunRecordingRule] {
        try await request(path: APIEndpoints.updateRecordingRule(id), method: "PUT", body: payload)
    }

    public func deleteRecordingRule(id: String) async throws -> [HDHomeRunRecordingRule] {
        try await request(path: APIEndpoints.deleteRecordingRule(id), method: "DELETE")
    }

    // MARK: - Live Watch APIs

    public func startWatch(channelNumber: String) async throws -> HDHomeRunRecording? {
        let resp: HDHomeRunRecording = try await request(path: APIEndpoints.startWatch(channelNumber: channelNumber), method: "POST")
        if resp.recordingId == nil, resp.sessionId == nil {
            return nil
        }
        return resp
    }

    public func heartbeatWatch(sessionId: String) async throws {
        _ = try await requestRaw(path: APIEndpoints.heartbeatWatch(sessionId: sessionId), method: "POST")
    }

    public func stopWatch(sessionId: String) async throws {
        _ = try await requestRaw(path: APIEndpoints.stopWatch(sessionId: sessionId), method: "POST")
    }

    public func promoteWatch(sessionId: String, options: [String: AnyCodable]? = nil) async throws -> HDHomeRunRecording {
        try await request(path: APIEndpoints.promoteWatch(sessionId: sessionId), method: "POST", body: options)
    }

    // MARK: - HLS Packaging APIs

    public func createChannelHLSSession(channelNumber: String, audioIndex: Int? = nil) async throws -> HDHomeRunRecording {
        try await request(path: APIEndpoints.hlsChannelSession(channelNumber: channelNumber, audioIndex: audioIndex), method: "POST")
    }

    public func createRecordingHLSSession(
        url: String,
        recordingId: String? = nil,
        start: Double? = nil,
        audioIndex: Int? = nil
    ) async throws -> HLSSessionResponse {
        struct RecordingStreamHLSBody: Encodable {
            let url: String
            let recordingId: String?
            let start: Double?
            let audioIndex: Int?

            enum CodingKeys: String, CodingKey {
                case url
                case recordingId = "recording_id"
                case start
                case audioIndex = "audio_index"
            }
        }
        return try await request(
            path: APIEndpoints.hlsRecordingSession(),
            method: "POST",
            body: RecordingStreamHLSBody(url: url, recordingId: recordingId, start: start, audioIndex: audioIndex)
        )
    }

    public func stopHLSSession(sessionId: String) async throws {
        _ = try await requestRaw(path: APIEndpoints.stopHLSSession(sessionId: sessionId), method: "POST")
    }

    // MARK: - Tuner APIs

    public func getTunerStatus() async throws -> [HDHomeRunTuner] {
        try await request(path: APIEndpoints.tunerStatus())
    }

    public func getTunerInfo() async throws -> HDHomeRunTunerInfo {
        try await request(path: APIEndpoints.tunerInfo())
    }

    // MARK: - Auth & Profile APIs

    public func listProfiles() async throws -> [UserProfile] {
        try await request(path: APIEndpoints.users())
    }

    public func login(userId: String, pin: String?, tokenName: String?) async throws -> CurrentUser {
        struct LoginBody: Codable {
            let pin: String?
            let tokenName: String?
            enum CodingKeys: String, CodingKey {
                case pin
                case tokenName = "token_name"
            }
        }
        return try await request(
            path: APIEndpoints.loginUser(userId),
            method: "POST",
            body: LoginBody(pin: pin, tokenName: tokenName)
        )
    }

    public func logout() async throws {
        _ = try await requestRaw(path: APIEndpoints.logoutUser(), method: "POST")
    }

    public func getCurrentUser() async throws -> CurrentUser {
        try await request(path: APIEndpoints.currentUser())
    }

    public func getUserPreferences() async throws -> UserPreferences {
        try await request(path: APIEndpoints.userPreferences())
    }

    public func updateUserPreferences(_ prefs: UserPreferences) async throws -> UserPreferences {
        try await request(path: APIEndpoints.userPreferences(), method: "PATCH", body: prefs)
    }

    // MARK: - Device APIs

    public func registerDevice() async throws -> DeviceRegisterResult {
        let result: DeviceRegisterResult = try await request(path: APIEndpoints.registerDevice(), method: "POST")
        deviceId = result.id
        return result
    }

    public func getCurrentDevice() async throws -> DeviceInfo {
        try await request(path: APIEndpoints.currentDevice())
    }

    public func listDevices() async throws -> [DeviceListEntry] {
        try await request(path: APIEndpoints.listDevices())
    }

    public func deleteDevice(id: String) async throws {
        _ = try await requestRaw(path: APIEndpoints.deleteDevice(id), method: "DELETE")
    }

    // MARK: - Settings APIs

    public func getSettings() async throws -> AppSettings {
        try await request(path: APIEndpoints.settings())
    }

    public func getTranscodePresets() async throws -> [HDHomeRunTranscodePreset] {
        try await request(path: APIEndpoints.transcodePresets())
    }

    public func getHWAccelDiagnostics() async throws -> HWAccelDiagnostics {
        try await request(path: APIEndpoints.hwaccelDiagnostics())
    }

    public func listNetworkIntegrations() async throws -> [NetworkIntegration] {
        try await request(path: APIEndpoints.networkIntegrations())
    }

    public func getNetworkIntegration(type: String) async throws -> NetworkIntegration {
        try await request(path: APIEndpoints.networkIntegration(type))
    }

    public func updateNetworkIntegration(type: String, settings: [String: AnyCodable]) async throws -> NetworkIntegration {
        try await request(path: APIEndpoints.networkIntegration(type), method: "PATCH", body: settings)
    }

    // MARK: - SyncPlay APIs

    public func createSyncPlayRoom(userName: String, initialContent: SyncPlayContent? = nil) async throws -> CreateSyncPlayRoomResponse {
        struct CreateRoomBody: Encodable {
            let userName: String
            let content: SyncPlayContent

            enum CodingKeys: String, CodingKey {
                case userName = "user_name"
                case content
            }
        }
        let fallbackContent = initialContent ?? SyncPlayContent(type: "channel", channelNumber: "default", title: "Live TV")
        let body = CreateRoomBody(userName: userName, content: fallbackContent)
        return try await request(path: APIEndpoints.syncPlayRooms(), method: "POST", body: body)
    }

    public func getSyncPlayRoom(code: String) async throws -> SyncPlayRoom {
        try await request(path: APIEndpoints.syncPlayRoom(code))
    }

    public func syncPlayWsUrl(roomCode: String, userName: String? = nil) -> URL? {
        // Note: the resulting URL carries the bearer token in its query string
        // (see APIEndpoints.syncPlayWs) - deliberately not logging the URL
        // itself here, only that construction failed and why.
        let path = APIEndpoints.syncPlayWs(roomCode: roomCode, userName: userName, token: bearerToken)
        guard let httpUrl = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            Log.network.error("syncPlayWsUrl: failed to construct base URL for room \(roomCode, privacy: .public)")
            return nil
        }
        var comps = URLComponents(url: httpUrl, resolvingAgainstBaseURL: true)
        if comps?.scheme == "https" {
            comps?.scheme = "wss"
        } else {
            comps?.scheme = "ws"
        }
        guard let result = comps?.url else {
            Log.network.error("syncPlayWsUrl: failed to resolve URLComponents for room \(roomCode, privacy: .public)")
            return nil
        }
        return result
    }
}
